nuvlaedge/agent/orchestrator/docker.py

```diff
@@ -608,43 +611,57 @@ def launch_job(self, job_id, job_execution_id, nuvla_endpoint,
                 Defaults to False.
             metrics (dict): A dictionary containing the metrics data which needs to be populated
-
         """
         cs = container_stats
-        cpu_percent = float('nan')
+        online_cpus = None
+        cpu_percent = None
+        container_id = metrics.get('id') or '?'
+        container_name = (metrics.get('name') or '?').lstrip('/')
+        precpu_stats = self.container_stats_cpu_prev.get(container_id)
+        container_msg = f'for container {container_id[:12]} ({container_name})'
 
         try:
             online_cpus_alt = len(cs["cpu_stats"]["cpu_usage"].get("percpu_usage", []))
             online_cpus = cs["cpu_stats"].get('online_cpus', online_cpus_alt)
 
-            cpu_delta = \
-                float(cs["cpu_stats"]["cpu_usage"]["total_usage"]) - \
-                float(cs["precpu_stats"]["cpu_usage"]["total_usage"])
-            system_delta = \
-                float(cs["cpu_stats"]["system_cpu_usage"]) - \
-                float(cs["precpu_stats"]["system_cpu_usage"])
-
-            if system_delta > 0.0:
-                cpu_percent = (cpu_delta / system_delta) * 100.0
+            if precpu_stats:
+                cpu_delta = \
+                    float(cs["cpu_stats"]["cpu_usage"]["total_usage"]) - \
+                    float(precpu_stats["cpu_usage"]["total_usage"])
+                system_delta = \
+                    float(cs["cpu_stats"]["system_cpu_usage"]) - \
+                    float(precpu_stats["system_cpu_usage"])
+
+                if cpu_delta >= 0.0 and system_delta > 0.0:
+                    cpu_percent = (cpu_delta / system_delta) * 100.0
+                else:
+                    logger.debug(f'Invalid CPU usage {container_msg}: '
+                                 f'({cpu_delta} / {system_delta}) * 100\n'
+                                 f'cpu_stats: {cs["cpu_stats"]}\n'
+                                 f'precpu_stats: {precpu_stats}')
+            else:
+                logger.debug(f'CPU usage not yet available {container_msg}')
         except (IndexError, KeyError, ValueError, ZeroDivisionError) as e:
-            logger.warning('Failed to get CPU usage for container '
-                           f'{cs.get("id", "?")[:12]} ({cs.get("name")}): {e}')
-            return
+            logger.warning(f'Failed to get CPU usage for container {container_msg}: {e}')
 
         if old_version:
-            metrics['cpu-percent'] = f'{round(cpu_percent):.2f}'
+            if cpu_percent is not None:
+                metrics['cpu-percent'] = f'{round(cpu_percent):.2f}'
         else:
-            metrics['cpu-usage'] = cpu_percent
-            metrics['cpu-capacity'] = online_cpus
+            if cpu_percent is not None:
+                metrics['cpu-usage'] = cpu_percent
+            if online_cpus is not None:
+                metrics['cpu-capacity'] = online_cpus
+
+        self.container_stats_cpu_prev[container_id] = cs.get('cpu_stats')
```

Comment:
I think this dictionary needs to be cleansed. Else, it will grow indefinitely. Probably after calling this function, all the container_id's not present in the iteration can be erased.